### PR TITLE
Txn pipeline

### PR DESCRIPTION
**Pull request by triump2020**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #4694 

## What this PR does / why we need it:
1. Finished implementing these interfaces: TxnStorage.Prepare, Committing, Commit, Rollback, PrecommitWrite, Close.
2. fixed some bugs 
3. remove useless code
TODO:
1.  Need to implement  interface : TxnStorage.StartRecovery ,  TxnStorage.Destroy.  These depends  on whether replay of TAE works correctly.
4. Add UT for 2PC

### CODE REVIEW

**Review by jiangxinmeng1 — APPROVED** [review]

File: pkg/vm/engine/tae/rpc/handle.go [hunk 1]
```diff
@@ -69,34 +74,45 @@ func (h *Handle) HandleRollback(meta txn.TxnMeta) (err error) {
...
 
 func (h *Handle) HandlePrepare(meta txn.TxnMeta) (pts timestamp.Timestamp, err error) {
-	txn, err := h.eng.GetTxnByID(meta.GetID())
+	var txn moengine.Txn
+	txn, err = h.eng.GetTxnByID(meta.GetID())
 	if err != nil {
 		return timestamp.Timestamp{}, err
 	}
-	ts, err := txn.Prepare()
+	participants := make([]uint64, len(meta.GetDNShards()))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/rpc/handle.go:93]: make(xxx, 0, len)

File: pkg/vm/engine/tae/txn/txnbase/command.go [hunk 2]
```diff
@@ -86,6 +91,12 @@ type TxnCmd struct {
 	Txn txnif.AsyncTxn
 }
 
+type TxnStateCmd struct {
+	ID    string
+	State txnif.TxnState
+	Cts   types.TS
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/txn/txnbase/command.go:97]: Change to CommitTs

File: pkg/vm/engine/tae/txn/txnbase/txn.go [hunk 3]
```diff
@@ -180,32 +179,42 @@ func (txn *Txn) Rollback() (err error) {
...
+		)
 	}
-	// TODO:
-	// Make committing log entry and flush and wait
 	if err = txn.ToCommittingFinished(); err != nil {
 		panic(err)
 	}
+	//Make a committing log entry, flush and wait it synced.
+	//A log entry's payload contains txn id , commit timestamp and txn's state.
+	_, err = txn.LogTxnState(true)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/txn/txnbase/txn.go:202]: We need to free log entry for reuse

File: pkg/vm/engine/tae/txn/txnbase/impl.go [hunk 4]
```diff
@@ -111,24 +114,32 @@ func (txn *Txn) commit2PC() (err error) {
...
 		if err = txn.ApplyCommit(); err != nil {
 			panic(err)
 		}
-		//TODO:Append committed log entry into log service asynchronously
-		//     for checkpointing the committing log entry
-		//txn.SetError(txn.LogTxnEntry())
 		txn.DoneWithErr(nil, false)
+		//Append a committed log entry into log service asynchronously
+		//     for checkpointing the committing log entry
+		_, err = txn.LogTxnState(false)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/txn/txnbase/impl.go:125]: 1. Free log entry when the log entry is flushed
2. We also need some mechanism to wait for async flush and handle any err

File: pkg/vm/engine/tae/txn/txnimpl/store.go [hunk 5]
```diff
@@ -113,6 +115,38 @@ func (store *txnStore) LogTxnEntry(dbId uint64, tableId uint64, entry txnif.TxnE
 	return db.LogTxnEntry(tableId, entry, readed)
 }
 
+func (store *txnStore) LogTxnState(sync bool) (logEntry entry.Entry, err error) {
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/txn/txnimpl/store.go:118]: LogTxnState is not a good name here

File: pkg/vm/engine/tae/txn/txnimpl/store.go [hunk 6]
```diff
@@ -113,6 +115,38 @@ func (store *txnStore) LogTxnEntry(dbId uint64, tableId uint64, entry txnif.TxnE
...
+	if err = logEntry.SetPayload(buf); err != nil {
+		return
+	}
+	info := &entry.Info{
+		Group: wal.GroupC,
+		TxnId: store.txn.GetID(),
+	}
+	logEntry.SetInfo(info)
+	var lsn uint64
+	lsn, err = store.driver.AppendEntry(wal.GroupC, logEntry)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/txn/txnimpl/store.go:139]: We need to maintain these LSNs in the txn workspace for further wal checkpoint

File: pkg/vm/engine/tae/txn/txnbase/command.go (hunk 2 above)
Comment (triump2020) [reply pkg/vm/engine/tae/txn/txnbase/command.go:97]: ok

**Review by XuPeng-SH — APPROVED** [review]
The next pr will address some of the mentioned issues

**Pull request merged**